### PR TITLE
INTIJC-5 - adding unit tests for AbstractConsumingSubStep

### PR DESCRIPTION
**Pull request by DMaxfield-BDS**
Initial add of unit tests for AbstractConsumingSubStep

### CODE REVIEW

File: src/test/java/com/synopsys/integration/stepworkflow/AbstractConsumingSubStepTest.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+            Arguments.of(false, failedData, null),
+            Arguments.of(false, failedData, runtimeException)
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("populateSuccessTests")
+    public void testAbstractConsumingSubStep_Success(boolean subStepSucceeded, String data, Exception e) {
+        AbstractConsumingSubStep<String> abstractConsumingSubStep = new AbstractConsumingSubStep_TestImpl();
+        SubStepResponse<String> subStepResponse_Input = new SubStepResponse<>(subStepSucceeded, data, e);
```
Comment (jamesrichard91) [inline src/test/java/com/synopsys/integration/stepworkflow/AbstractConsumingSubStepTest.java:50]: Should `SubStepResponse` be updated to have 2 static methods to construct it, and make the constructor private? One for success and one for failure? It is odd that when things are successful that there is an exception.

File: src/test/java/com/synopsys/integration/stepworkflow/AbstractConsumingSubStepTest.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+public class AbstractConsumingSubStepTest {
+
+    private final String verificationData = "Verification output";
+    private final RuntimeException verificationException = new RuntimeException("Verification runtime exception");
+    private final SubStepResponse<Object> subStepResponse_successOutput = new SubStepResponse<>(true, verificationData, verificationException);
```
Comment (rottebds) [inline src/test/java/com/synopsys/integration/stepworkflow/AbstractConsumingSubStepTest.java:18]: ConsumingSubSteps should not return data. While testing that it does is technically a valid code path, that is not the intended use of the class and testing it conveys an incorrect API to users of this class. I should consider preventing this in the API of AbstractConsumingSubStep.

File: src/test/java/com/synopsys/integration/stepworkflow/AbstractConsumingSubStepTest.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+    private static Stream<Arguments> populateSuccessTests() {
+        final String successData = "Success Data";
+        RuntimeException runtimeException = new RuntimeException("Success runtime exception");
+        return Stream.of(
```
Comment (rottebds) [inline src/test/java/com/synopsys/integration/stepworkflow/AbstractConsumingSubStepTest.java:23]: These arguments appear to be for constructing SubStepResponses-- why not just provide a stream of SubStepResponses to test against?

File: src/test/java/com/synopsys/integration/stepworkflow/AbstractConsumingSubStepTest.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+            Arguments.of(false, failedData, runtimeException)
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("populateSuccessTests")
+    public void testAbstractConsumingSubStep_Success(boolean subStepSucceeded, String data, Exception e) {
+        AbstractConsumingSubStep<String> abstractConsumingSubStep = new AbstractConsumingSubStep_TestImpl();
+        SubStepResponse<String> subStepResponse_Input = new SubStepResponse<>(subStepSucceeded, data, e);
+        SubStepResponse<Object> subStepResponse_Output = abstractConsumingSubStep.run(subStepResponse_Input);
```
Comment (rottebds) [inline src/test/java/com/synopsys/integration/stepworkflow/AbstractConsumingSubStepTest.java:51]: For whatever it's worth, underscores in non-enum/static-final variable names and in method names at all is a discouraged practice: https://docs.oracle.com/javase/tutorial/java/nutsandbolts/variables.html

previousSubStepResponse could replace subStepResponse_Input here and abstractConsumingSubStepResponse could replace subStepResponse_Output here-- it'd probably convey the api of the class better.

**Comment by rottebds** [comment]
Most of my comments here apply to #10 and #11 as well-- to avoid repeating myself, I'll avoid commenting on those PRs until this one is squared away.

File: src/test/java/com/synopsys/integration/stepworkflow/AbstractConsumingSubStepTest.java (hunk 2 above)
Comment (DMaxfield-BDS) [reply src/test/java/com/synopsys/integration/stepworkflow/AbstractConsumingSubStepTest.java:18]: @rottebds So with the static constructors, I would not be able to test having a SubStepResponse who's subStepSucceeded == true AND has an exception. Of course when subStepSucceeded == true it shouldn't have an exception, but this is allowed with the constructor. If using the static constructors is the intended usage AND the true+exception I mentioned above isn't needed, I will update this asap.

File: src/test/java/com/synopsys/integration/stepworkflow/AbstractConsumingSubStepTest.java (hunk 3 above)
Comment (DMaxfield-BDS) [reply src/test/java/com/synopsys/integration/stepworkflow/AbstractConsumingSubStepTest.java:23]: Purely to help anyone debugging failed tests. When data points are passed, the name of the test within the output html will be something like:

```
[1] true, , null
[2] true, , java.lang.RuntimeException: Success runtime exception
```
When passing in an object, the name of the test will be something like:

```
[1] com.synopsys.integration.stepworkflow.SubStepResponse@59816976
[2] com.synopsys.integration.stepworkflow.SubStepResponse@3be33474
```

If the benefit of displaying the data points in the html is not greater than what it detracts from the code, I will update this asap as well.

Comment (rottebds) [reply src/test/java/com/synopsys/integration/stepworkflow/AbstractConsumingSubStepTest.java:23]: Data in HTML sounds compelling to me.

File: src/test/java/com/synopsys/integration/stepworkflow/AbstractConsumingSubStepTest.java (hunk 2 above)
Comment (rottebds) [reply src/test/java/com/synopsys/integration/stepworkflow/AbstractConsumingSubStepTest.java]: @DMaxfield-BDS I retract my previous statement-- the tests as they are now are the best way to test the class. If the class gets updated in the future to make constructing a SubStepResponse "improperly" difficult, then the tests should be updated to reflect that. However, until then these tests are good as-is.

**Pull request merged**